crates/uv-globfilter/src/glob_dir_filter.rs

```diff
@@ -0,0 +1,277 @@
+use bstr::{ByteSlice, ByteVec};
+use globset::{Glob, GlobSet, GlobSetBuilder};
+use regex_automata::dfa;
+use regex_automata::dfa::Automaton;
+use std::path::{Path, MAIN_SEPARATOR, MAIN_SEPARATOR_STR};
+use tracing::warn;
+
+/// Chosen at a whim -Konsti
+const DFA_SIZE_LIMIT: usize = 1_000_000;
+
+/// Filter a directory tree traversal (walkdir) by whether any paths of a directory can be included
+/// at all.
+///
+/// Internally, the globs are converted to a regex and then to a DFA, which unlike the globs and the
+/// regex allows to check for prefix matches.
+pub struct GlobDirFilter {
+    glob_set: GlobSet,
+    dfa: Option<dfa::dense::DFA<Vec<u32>>>,
+}
+
+impl GlobDirFilter {
+    /// The filter matches if any of the globs matches.
+    ///
+    /// See <https://github.com/BurntSushi/ripgrep/discussions/2927> for the error returned.
+    pub fn from_globs(globs: &[Glob]) -> Result<Self, globset::Error> {
+        let mut glob_set_builder = GlobSetBuilder::new();
+        for glob in globs {
+            glob_set_builder.add(glob.clone());
+        }
+        let glob_set = glob_set_builder.build()?;
+
+        let regexes: Vec<_> = globs
+            .iter()
+            .map(|glob| {
+                let main_separator = regex::escape(MAIN_SEPARATOR_STR);
+                let regex = glob
+                    .regex()
+                    // We are using a custom DFA builder
+                    .strip_prefix("(?-u)")
+                    .expect("a glob is a non-unicode byte regex")
+                    // Match windows paths if applicable
+                    .replace('/', &main_separator);
```

Comment:
This is a different strategy than globset, it avoids normalizing the path itself.

Reply:
Hmmm, what if the path has both a `/` and a `\` in it?

Reply:
Good question, can we reasonably support this? Given that we want to (need to) be cross platform, should we error if we ever encounter any path with a `/` (windows) or a `\` (linux) inside a component?

Reply:
Oh hmmm, not within a component. What I mean is that I've seen Windows paths where `\` and `/` are inter-mingled. For example, `C:\Users\andrew\home/bin/whatever`. Or at least, I think I have... So in that circumstance, `\` and `/` would both be a separator.